File: Nearest Exit From Entrance In Maze.cpp
class Solution {
public:
    int nearestExit(vector<vector<char>> &maze, vector<int> &entrance) {
        int dx[] = {1, -1, 0, 0};
        int dy[] = {0, 0, 1, -1};
        int n = (int) maze.size();
        int m = (int) maze[0].size();
        int visited[n][m];
        memset(visited, 0, sizeof(visited));
        queue<vector<int>> q;
        q.push({entrance[0], entrance[1], 0});
        visited[entrance[0]][entrance[1]] = 1;

        while (!q.empty()) {
            vector<int> cur = q.front();
            q.pop();
            if ((cur[0] == 0 || cur[1] == 0 || cur[0] == n - 1 || cur[1] == m - 1) &&
                !(cur[0] == entrance[0] && cur[1] == entrance[1])) {
                return cur[2];
            }
            for (int i = 0; i < 4; i++) {
                int new_x = dx[i] + cur[0];
                int new_y = dy[i] + cur[1];
                if (new_x >= 0 && new_x < n && new_y >= 0 && new_y < m && maze[new_x][new_y] == '.' &&
                    !visited[new_x][new_y]) {
                    q.push({new_x, new_y, cur[2] + 1});
                    visited[new_x][new_y] = 1;
                }
            }
        }
        return -1;
    }
};

File: Nearest Exit From Entrance In Maze.java
class Solution {
    public static class Cell {
        public int u;
        public int v;
        public int steps;

        public Cell(int u, int v, int lvl) {
            super();
            this.u = u;
            this.v = v;
            this.steps = lvl;
        }

    }

    public int nearestExit(char[][] maze, int[] entrance) {
        int m = maze.length;
        int n = maze[0].length;
        boolean[][] vis = new boolean[m][n];
        int[] dx = { -1, 0, 0, 1 };
        int[] dy = { 0, -1, 1, 0 };

        Queue<Cell> queue = new LinkedList<Cell>();
        int result = Integer.MAX_VALUE;
        int steps = 0;
        Cell currCell = new Cell(entrance[0], entrance[1], steps);
        queue.add(currCell);
        while (!queue.isEmpty()) {
            currCell = queue.poll();
            steps = currCell.steps;
            int u = currCell.u;
            int v = currCell.v;
            if ((u == 0 || u == m - 1 || v == 0 || v == n - 1) && (u != entrance[0] || v != entrance[1])) {
                result = Math.min(result, steps);
            }
            if (!vis[u][v]) {
                vis[u][v] = true;
                for (int i = 0; i < 4; i++) {
                    int x = u + dx[i];
                    int y = v + dy[i];
                    if (x >= 0 && y >= 0 && x < m && y < n && !vis[x][y] && maze[x][y] == '.') {
                        queue.add(new Cell(x, y, steps + 1));
                    }
                }
            }
        }
        return result == Integer.MAX_VALUE ? -1 : result;
    }
}
